packages/protocol/lib/compatibility/verify-bytecode.ts

```diff
@@ -0,0 +1,277 @@
+  const relevantTx = context.proposal.find((tx: ProposalTx) => {
+    return tx.contract === `${contract}Proxy`
+  })
+
+  return relevantTx.args[0]
+}
+
+const getSoureBytecode = (contract: string, context: VerificationContext): string => {
+  const sourceBytecodeWithMetadata = context.artifacts.getArtifactByName(contract).deployedBytecode
+  return stripMetadata(sourceBytecodeWithMetadata)
+}
+
+const getLibraryAddress = async (contract: string, context: VerificationContext): Promise<string> => {
+  const proxyAddress = context.libraryAddresses[contract]
+  await verifyProxy(proxyAddress, context)
+
+  if (isImplementationChanged(contract, context)) {
+    return getProposedImplementationAddress(contract, context)
+  } else {
+    return getProxiedAddress(proxyAddress, context)
+  }
+}
+
+const getCoreContractAddress = async (contract: string, context: VerificationContext): Promise<string> => {
+  const proxyAddress: string = isProxyChanged(contract, context) ?
+    getProposedProxyAddress(contract, context) :
+    await getRegisteredProxyAddress(contract, context)
+
+  await verifyProxy(proxyAddress, context)
+
+  if (isImplementationChanged(contract, context)) {
+    return getProposedImplementationAddress(contract, context)
+  } else {
+    return getProxiedAddress(proxyAddress, context)
+  }
+}
+
+const getImplementationAddress = async (contract: string, isLibrary: boolean, context: VerificationContext): Promise<string> => {
+  // Where we find the implementation address depends on two factors:
+  // 1. Is the contract a core contract registered in the Registry vs. a linked
+  // library.
+  // 2. Is the contract affected by a governance proposal.
+  // TODO: remove isBeforeRelease1. Before the first contracts upgrade,
+  // libraries are not proxied.
+  if (isLibrary && context.isBeforeRelease1) {
+    return '0x' + context.libraryAddresses[contract]
+  } else if (isLibrary) {
+    return getLibraryAddress(contract, context)
+  } else {
+    return getCoreContractAddress(contract, context)
+  }
+
+}
+
+const getOnchainBytecode = async (contract: string, isLibrary: boolean, context: VerificationContext): Promise<string> => {
+  const implementationAddress = await getImplementationAddress(contract, isLibrary, context)
+  const onchainBytecodeWithMetadata = await context.web3.eth.getCode(implementationAddress)
+  return stripMetadata(onchainBytecodeWithMetadata)
+}
+
+const dfsStep = async (queue: string[], visited: Set<string>, context: VerificationContext) => {
+  const contract = queue.pop()
+  if (ignoredContracts.includes(contract)) {
+    return
+  }
+
+  const isLibrary = !context.contracts.includes(contract)
+
+  const sourceBytecode = getSoureBytecode(contract, context)
+  const onchainBytecode = await getOnchainBytecode(contract, isLibrary, context)
+
+  const sourceLibraryPositions = collectLibraryPositions(sourceBytecode)
+  collectLibraryAddresses(onchainBytecode, sourceLibraryPositions, context.libraryAddresses)
```

Comment:
Would you say that making `LibraryAddresses` a class, and `collect` a method on it, makes it clear enough that this could mutate?

I think in future releases we could modify this to not even be a DFS. With libraries being proxied, looks like Asa is having even them be entered in the Registry. At that point we can just feed this script a list of all core contracts and libraries, and like you're suggesting, gather addresses in a preprocessing step, using the Registry as a source of truth for that.

I suggest we leave this as a future improvement.